File: src/org/wavescale/sourcesync/synchronizer/SCPFileSynchronizer.java
package org.wavescale.sourcesync.synchronizer;

import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import com.jcraft.jsch.*;
import org.jetbrains.annotations.NotNull;
import org.wavescale.sourcesync.api.FileSynchronizer;
import org.wavescale.sourcesync.api.Utils;
import org.wavescale.sourcesync.config.SCPConfiguration;
import org.wavescale.sourcesync.logger.BalloonLogger;
import org.wavescale.sourcesync.logger.EventDataLogger;

import java.io.*;
import java.nio.file.Paths;

/**
 * ****************************************************************************
 * Copyright (c) 2005-2014 Faur Ioan-Aurel.                                     *
 * All rights reserved. This program and the accompanying materials             *
 * are made available under the terms of the MIT License                        *
 * which accompanies this distribution, and is available at                     *
 * http://opensource.org/licenses/MIT                                           *
 * *
 * For any issues or questions send an email at: fioan89@gmail.com              *
 * *****************************************************************************
 */
public class SCPFileSynchronizer extends FileSynchronizer {
    public static final String SSH_KNOWN_HOSTS = Paths.get(System.getProperty("user.home"), ".ssh", "known_hosts").toString();
    private JSch jsch;
    private Session session;


    public SCPFileSynchronizer(@NotNull SCPConfiguration connectionInfo, @NotNull Project project, @NotNull ProgressIndicator indicator) {
        super(connectionInfo, project, indicator);
        this.jsch = new JSch();
        this.getIndicator().setIndeterminate(true);
    }

    @Override
    public boolean connect() {
        if (!isConnected()) {
            try {
                initSession();
                session.setPassword(this.getConnectionInfo().getUserPassword());
                this.session.setConfig("StrictHostKeyChecking", "no");
                this.session.connect();
                this.setConnected(true);
                return true;
            } catch (JSchException e) {
                EventDataLogger.logWarning(e.toString(), this.getProject());
                return false;
            }
        }
        return true;
    }

    private void initSession() throws JSchException {
        SCPConfiguration configuration = (SCPConfiguration) this.getConnectionInfo();
        session = this.jsch.getSession(this.getConnectionInfo().getUserName(), this.getConnectionInfo().getHost(),
                this.getConnectionInfo().getPort());
        if (configuration.isPasswordlessSSHSelected()) {
            session.setConfig("PreferredAuthentications", "publickey");
            try {
                Utils.createFile(SSH_KNOWN_HOSTS);
            } catch (IOException e) {
                EventDataLogger.logError("Could not identify nor create the ssh known hosts file at " + SSH_KNOWN_HOSTS + ". The returned error is:" + e.getMessage(), this.getProject());
            }
            this.jsch.setKnownHosts(SSH_KNOWN_HOSTS);
            // add private key and passphrase if exists
            if (configuration.isPasswordlessWithPassphrase()) {
                this.jsch.addIdentity(configuration.getCertificatePath(), configuration.getUserPassword());
            } else {
                this.jsch.addIdentity(configuration.getCertificatePath());
            }

        } else {
            session.setPassword(this.getConnectionInfo().getUserPassword());
        }
    }

    @Override
    public void disconnect() {
        if (this.session != null) {
            this.session.disconnect();
            this.setConnected(false);
        }
    }

    /**
     * Uploads the given file to the remote target.
     *
     * @param sourcePath      a <code>String</code> representing a file path to be uploaded. This is a relative path
     *                        to project base path.
     * @param destinationPath a <code>String</code> representing a location path on the remote target
     *                        where the source will be uploaded.
     */
    @Override
    public void syncFile(String sourcePath, String destinationPath) {
        boolean preserveTimestamp = this.getConnectionInfo().isPreserveTime();
        // exec 'scp -t rfile' remotely
        String finalSourcePath = Utils.buildUnixPath(getProject().getBasePath(), sourcePath);
        String remotePath = Utils.buildUnixPath(this.getConnectionInfo().getRootPath(), destinationPath);

        try {
            String command = "scp " + (preserveTimestamp ? "-p" : "") + " -t -C " + remotePath;
            Channel channel = this.session.openChannel("exec");
            ((ChannelExec) channel).setCommand(command);

            // get I/O streams for remote scp
            OutputStream out = channel.getOutputStream();
            InputStream in = channel.getInputStream();

            channel.connect();

            if (checkAck(in) != 0) {
                return;
            }

            File _lfile = new File(finalSourcePath);
            this.getIndicator().setIndeterminate(false);
            this.getIndicator().setText("Uploading...[" + _lfile.getName() + "]");
            if (preserveTimestamp) {
                command = "T " + (_lfile.lastModified() / 1000) + " 0";
                // The access time should be sent here,
                // but it is not accessible with JavaAPI ;-<
                command += (" " + (_lfile.lastModified() / 1000) + " 0\n");
                out.write(command.getBytes());
                out.flush();
                if (checkAck(in) != 0) {
                    return;
                }
            }
            // send "C0644 filesize filename", where filename should not include '/'
            long filesize = _lfile.length();
            command = "C0644 " + filesize + " ";
            if (finalSourcePath.lastIndexOf('/') > 0) {
                command += finalSourcePath.substring(finalSourcePath.lastIndexOf('/') + 1);
            } else {
                command += finalSourcePath;
            }
            command += "\n";
            out.write(command.getBytes());
            out.flush();
            if (checkAck(in) != 0) {
                return;
            }

            // send content of finalSourcePath
            FileInputStream fis = new FileInputStream(finalSourcePath);
            double totalUploaded = 0.0;
            byte[] buf = new byte[1024];
            while (true) {
                int len = fis.read(buf, 0, buf.length);
                if (len <= 0) break;
                out.write(buf, 0, len); //out.flush();
                totalUploaded += len;
                this.getIndicator().setFraction(totalUploaded / filesize);
            }
            fis.close();
            // send '\0'
            buf[0] = 0;
            out.write(buf, 0, 1);
            out.flush();
            if (checkAck(in) != 0) {
                return;
            }
            out.close();
            channel.disconnect();
        } catch (IOException e) {
            EventDataLogger.logWarning(e.toString(), getProject());
        } catch (JSchException e) {
            EventDataLogger.logWarning(e.toString(), getProject());
        }

    }

    private int checkAck(InputStream in) throws IOException {
        int b = in.read();
        // b may be 0 for success,
        // 1 for error,
        // 2 for fatal error,
        // -1
        if (b == 0) return b;
        if (b == -1) return b;

        if (b == 1 || b == 2) {
            StringBuilder sb = new StringBuilder();
            int c;
            do {
                c = in.read();
                sb.append((char) c);
            }
            while (c != '\n');
            if (b == 1) { // error
                BalloonLogger.logBalloonError(sb.toString(), this.getProject());
                EventDataLogger.logError(sb.toString(), getProject());
            }
            if (b == 2) { // fatal error
                BalloonLogger.logBalloonError(sb.toString(), this.getProject());
                EventDataLogger.logError(sb.toString(), getProject());
            }
        }
        return b;
    }
}

File: src/main/kotlin/org/wavescale/sourcesync/synchronizer/SCPFileSynchronizer.kt
/**
 * ****************************************************************************
 * Copyright (c) 2014-2022 Faur Ioan-Aurel.                                     *
 * All rights reserved. This program and the accompanying materials             *
 * are made available under the terms of the MIT License                        *
 * which accompanies this distribution, and is available at                     *
 * http://opensource.org/licenses/MIT                                           *
 * *
 * For any issues or questions send an email at: fioan89@gmail.com              *
 * *****************************************************************************
 */
package org.wavescale.sourcesync.synchronizer

import com.intellij.openapi.components.service
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.jcraft.jsch.ChannelExec
import com.jcraft.jsch.JSch
import com.jcraft.jsch.JSchException
import com.jcraft.jsch.Session
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.nio.file.Path
import java.nio.file.Paths
import org.wavescale.sourcesync.SourcesyncBundle
import org.wavescale.sourcesync.api.Utils
import org.wavescale.sourcesync.configurations.AuthenticationType
import org.wavescale.sourcesync.configurations.ScpSyncConfiguration
import org.wavescale.sourcesync.notifications.Notifier
import org.wavescale.sourcesync.services.StatsService
import org.wavescale.sourcesync.services.SyncStatusService

class SCPFileSynchronizer(private val configuration: ScpSyncConfiguration, val project: Project) : Synchronizer {
    private val syncStatusService = service<SyncStatusService>()
    private val statsService = service<StatsService>()
    private val jsch: JSch = JSch()
    private var session: Session? = null

    private var isConnected: Boolean = false

    override fun connect(): Boolean {
        return if (!isConnected) {
            try {
                initSession()
                session!!.connect()
                isConnected = true
                true
            } catch (e: JSchException) {
                syncStatusService.removeRunningSync(configuration.name)
                Notifier.notifyError(
                    project,
                    SourcesyncBundle.message("scp.upload.fail.title"),
                    "Can't open SCP connection to ${configuration.hostname}. Reason: ${e.message}",
                )
                false
            }
        } else true
    }

    @Throws(JSchException::class)
    private fun initSession() {
        syncStatusService.addRunningSync(configuration.name)
        session = jsch.getSession(
            configuration.username, configuration.hostname,
            configuration.port.toInt()
        )
        session!!.setConfig("StrictHostKeyChecking", "no")
        if (configuration.authenticationType == AuthenticationType.KEY_PAIR) {
            session!!.setConfig("PreferredAuthentications", "publickey")
            try {
                Utils.createFile(SSH_KNOWN_HOSTS)
            } catch (e: IOException) {
                syncStatusService.removeRunningSync(configuration.name)
                Notifier.notifyError(
                    project,
                    SourcesyncBundle.message("scp.upload.fail.title"),
                    "Could not identify nor create the SSH known hosts file at $SSH_KNOWN_HOSTS. Reason: ${e.message}",
                )
                return
            }
            jsch.setKnownHosts(SSH_KNOWN_HOSTS)
            // add private key and passphrase if exists
            if (configuration.passphrase.isNullOrEmpty().not()) {
                jsch.addIdentity(configuration.privateKey, configuration.passphrase)
            } else {
                jsch.addIdentity(configuration.privateKey)
            }
        } else {
            session!!.setPassword(configuration.password)
        }
    }

    override fun disconnect() {
        try {
            session?.disconnect()
        } finally {
            isConnected = false
            syncStatusService.removeRunningSync(configuration.name)
        }
    }

    /**
     * Uploads the given file to the remote target.
     *
     * @param src     a `String` representing a file path to be uploaded. This is a relative path
     * to project base path.
     * @param uploadLocation a `String` representing a location path on the remote target
     * where the source will be uploaded.
     */
    override fun syncFile(src: String, uploadLocation: Path, indicator: ProgressIndicator) {
        val srcAsFile = File(src)
        var remotePath = Paths
            .get(configuration.workspaceBasePath)
            .resolve(uploadLocation)
            .pathStringLike(configuration.workspaceBasePath)

        var command = "scp " + (if (configuration.preserveTimestamps) "-p" else "") + " -t -C "
        if (srcAsFile.isFile) {
            remotePath = Paths
                .get(remotePath)
                .resolve(srcAsFile.name)
                .pathStringLike(configuration.workspaceBasePath)
            command += remotePath
        }
        try {
            val channel = session!!.openChannel("exec")
            (channel as ChannelExec).setCommand(command)

            // get I/O streams for remote scp
            val out = channel.getOutputStream()
            val inputStream = channel.getInputStream()
            channel.connect()
            if (checkAck(inputStream, this::onChannelConnectError) != 0) {
                return
            }
            indicator.isIndeterminate = false
            indicator.text = "Uploading...[" + srcAsFile.name + "]"
            if (configuration.preserveTimestamps) {
                command = "T " + srcAsFile.lastModified() / 1000 + " 0"
                // The access time should be sent here,
                // but it is not accessible with JavaAPI ;-<
                command += " " + (srcAsFile.lastModified() / 1000) + " 0\n"
                out.write(command.toByteArray())
                out.flush()
                if (checkAck(inputStream, this::onPreservingTimestampsError) != 0) {
                    return
                }
            }
            // send "C0644 filesize filename", where filename should not include '/'
            val filesize = srcAsFile.length()
            command = "C0644 $filesize "
            command += Paths.get(src).fileName.toString()
            command += "\n"
            out.write(command.toByteArray())
            out.flush()
            if (checkAck(inputStream, this::onSendingFileModesError) != 0) {
                return
            }

            // send content of finalSourcePath
            val fis = FileInputStream(src)
            var totalUploaded = 0.0
            val buf = ByteArray(1024)
            while (true) {
                val len = fis.read(buf, 0, buf.size)
                if (len <= 0) break
                out.write(buf, 0, len) //out.flush();
                totalUploaded += len.toDouble()
                indicator.fraction = totalUploaded / filesize
            }
            fis.close()
            // send '\0'
            buf[0] = 0
            out.write(buf, 0, 1)
            out.flush()
            if (checkAck(inputStream, this::onSendFileContentError) != 0) {
                return
            }
            out.close()
            channel.disconnect()
            statsService.registerSuccessfulUpload()
            if (statsService.eligibleForDonations()) {
                Notifier.notifyDonation(project)
            }
        } catch (e: Exception) {
            syncStatusService.removeRunningSync(configuration.name)
            Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                "Upload to ${configuration.hostname} failed. Reason: ${e.message}",
            )
        }
    }

    private fun onChannelConnectError(errorCode: AckError, reason: String) {
        when (errorCode) {
            AckError.ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.channel.connect.error.message", configuration.hostname, reason)
            )

            AckError.FATAL_ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.channel.connect.fatal.error.message", configuration.hostname, reason)
            )

            AckError.UNKNOWN -> Unit
        }
    }

    private fun onPreservingTimestampsError(errorCode: AckError, reason: String) {
        when (errorCode) {
            AckError.ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.preserve.timestamps.error.message", reason)
            )

            AckError.FATAL_ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.preserve.timestamps.fatal.error.message", reason)
            )

            AckError.UNKNOWN -> Unit
        }
    }

    private fun onSendingFileModesError(errorCode: AckError, reason: String) {
        when (errorCode) {
            AckError.ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.file.mode.error.message", configuration.hostname, reason)
            )

            AckError.FATAL_ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.file.mode.fatal.error.message", configuration.hostname, reason)
            )

            AckError.UNKNOWN -> Unit
        }
    }

    private fun onSendFileContentError(errorCode: AckError, reason: String) {
        when (errorCode) {
            AckError.ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.file.content.error.message", configuration.hostname, reason)
            )

            AckError.FATAL_ERROR -> Notifier.notifyError(
                project,
                SourcesyncBundle.message("scp.upload.fail.title"),
                SourcesyncBundle.message("scp.upload.fail.file.content.fatal.error.message", configuration.hostname, reason)
            )

            AckError.UNKNOWN -> Unit
        }
    }

    @Throws(IOException::class)
    /**
     * Reads the server response (ack) for a client command
     */
    private fun checkAck(inStream: InputStream, onError: (errorCode: AckError, reason: String) -> Unit): Int {
        val b = inStream.read()
        // b may be 0 for success,
        // 1 for error,
        // 2 for fatal error,
        // -1
        if (b == 0) return b
        if (b == -1) return b

        val sb = StringBuilder()
        var c: Int
        do {
            c = inStream.read()
            sb.append(c.toChar())
        } while (c != '\n'.code)
        syncStatusService.removeRunningSync(configuration.name)
        onError(AckError.from(b), sb.toString())
        return b
    }

    private enum class AckError(private val code: Int) {
        ERROR(1), FATAL_ERROR(2), UNKNOWN(3);

        companion object {
            infix fun from(value: Int) = AckError.values().firstOrNull { it.code == value } ?: UNKNOWN
        }
    }

    companion object {
        val SSH_KNOWN_HOSTS = Paths.get(System.getProperty("user.home"), ".ssh", "known_hosts").toString()
    }
}